crates/nu-cli/src/commands/math/ceil.rs

```diff
@@ -0,0 +1,98 @@
+use crate::commands::math::utils::run_with_function;
+use crate::commands::WholeStreamCommand;
+use crate::prelude::*;
+use bigdecimal::One;
+use nu_errors::ShellError;
+use nu_protocol::{Primitive, Signature, UntaggedValue, Value};
+
+pub struct SubCommand;
+
+#[async_trait]
+impl WholeStreamCommand for SubCommand {
+    fn name(&self) -> &str {
+        "math ceil"
+    }
+
+    fn signature(&self) -> Signature {
+        Signature::build("math celi")
+    }
+
+    fn usage(&self) -> &str {
+        "Applies the ceil function to a list of numbers"
+    }
+
+    async fn run(
+        &self,
+        args: CommandArgs,
+        registry: &CommandRegistry,
+    ) -> Result<OutputStream, ShellError> {
+        run_with_function(
+            RunnableContext {
+                input: args.input,
+                registry: registry.clone(),
+                shell_manager: args.shell_manager,
+                host: args.host,
+                ctrl_c: args.ctrl_c,
+                current_errors: args.current_errors,
+                name: args.call_info.name_tag,
+                raw_input: args.raw_input,
+            },
+            ceil,
+        )
+        .await
+    }
+
+    fn examples(&self) -> Vec<Example> {
+        vec![Example {
+            description: "Apply the ceil function to a list of numbers",
+            example: "echo [1.5 2.3 -3.1] | math ceil",
+            result: Some(vec![
+                UntaggedValue::int(2).into(),
+                UntaggedValue::int(3).into(),
+                UntaggedValue::int(-3).into(),
+            ]),
+        }]
+    }
+}
+
+fn ceil_big_decimal(val: &BigDecimal) -> BigInt {
+    let mut maybe_ceiled = val.round(0);
+    if &maybe_ceiled < val {
+        maybe_ceiled += BigDecimal::one();
+    }
+    let (ceiled, _) = maybe_ceiled.into_bigint_and_exponent();
+    ceiled
+}
+
+/// Calculate product of given values
+pub fn ceil(values: &[Value], _name: &Tag) -> Result<Value, ShellError> {
+    if !values.iter()
```

Comment:
This looks good generally. I think the one place that stood out is that you do a check here to make sure all values are an allowed type. For streams that might be very long, or not have an end, this wouldn't terminate.

Instead, you may just want to detect if a value doesn't match at the same time you're doing the mapping over the input stream. If you find something that isn't the right type, emit the error then.

Reply:
Correct me if I'm wrong, but `values` is a slice and therefore arbitrary large but finite. So this will always terminate.

But it's no problem to move the detection down to the match where I apply round/floor/ceil.

Reply:
Right, instead of passing in a slice here, operate on the InputStream itself. You can use functions like`.map` to apply a function to the stream of data as it passes. For this PR, that function could pattern match the values, and apply the ceil/floor/round functionality to the numeric ones.